### PR TITLE
[ENG-4036][ENG-4037] Spam extraction logic for NotableDomains

### PR DESCRIPTION
**Pull request by Johnetordoff**
## QA Notes

Please make verification statements inspired by your code and what your code touches.
- Verify
- Verify

What are the areas of risk?

Any concerns/considerations/questions that development raised?

## Ticket

https://openscience.atlassian.net/browse/ENG-4037

### CODE REVIEW

File: osf/models/notable_domain.py [hunk 1]
```diff
@@ -40,6 +45,56 @@ def __repr__(self):
     def __str__(self):
         return repr(self)
 
+    @staticmethod
+    def check_resource_for_domains(obj, confirm_spam=False, send_to_moderation=False):
```
Comment (adlius) [inline osf/models/notable_domain.py:49]: I *think* we might be putting a bit more logic than necessary here in this static method because it needs to have a reference to the `obj` on which we extract domains. This is kinda antithetical to the idea of a static method, which shouldn't rely on data of a specific instance, but only on the relevant data of a class.

How about we slim down this static method a bit so that it only takes `content` and `guid` as parameters and `get_or_create` instances for `NotableDomain` and `DomainReference`? If we go this route, then we can define another `extract_domains` method in the `SpamMixin` that invokes this static method. The benefit of this solution is that a class method is always self-aware, so that we don't have to pass an `obj` around. We may also override the `extract_domains` behavior for each subclass of `SpamMixin` to have more granular control over domain extraction behavior.

File: osf/models/notable_domain.py [hunk 2]
```diff
@@ -40,6 +45,56 @@ def __repr__(self):
...
+                defaults={'note': NotableDomain.Note.UNKNOWN}
+            )
+            DomainReference.objects.get_or_create(
+                domain=notable_domain,
+                referrer_object_id=obj.id,
+                referrer_content_type=ContentType.objects.get_for_model(obj)
+            )
+
+    @staticmethod
+    def has_spam_domain(obj):
```
Comment (adlius) [inline osf/models/notable_domain.py:88]: I think this should also be a method on the `SpamMixin` class because it is describing whether a resource have contents that contains spam domains.

We could modify this static method to a `content_contains_spam_domain` method that takes in only `content` and returns a boolean. And then we can have the `has_spam_domain` method on the `SpamMixin` to call this static method.

File: osf/models/notable_domain.py [hunk 3]
```diff
@@ -40,6 +45,56 @@ def __repr__(self):
...
+        for field in list(obj.SPAM_CHECK_FIELDS):
+            fields_to_query += [models.Q(**{f'{field}__regex': settings.DOMAIN_REGEX})]
+        query = reduce(operator.or_, fields_to_query)
+        has_domain = obj.__class__.objects.filter(query).filter(id=obj.id).exists()
+
+        if not has_domain:
+            return False
+        elif confirm_spam and NotableDomain.has_spam_domain(obj):
+            obj.confirm_spam(save=True)
+        elif send_to_moderation:
```
Comment (adlius) [inline osf/models/notable_domain.py:60]: Do we really need the `send_to_moderation` flag? If so this is more reason why this should be on the `SpamMixin` rather than the `NotableDomain` class.

File: osf/models/spam.py [hunk 4]
```diff
@@ -221,6 +224,8 @@ def do_check_spam(self, author, author_email, content, request_headers, update=T
         if self.is_spammy:
             return True
 
+        enqueue_task(check_resource_for_domains.s(guid=self.guids.first()._id))
```
[the remark this answers is not in the document]
Comment (Johnetordoff) [reply osf/models/spam.py:227]: > If we only extract the domains here, that means it will never extract domains for spammy or hammy resources. 

Yes this is expected behavior, once we've backfilled everything there's no need to recheck Spammed or Hammed resources.

> Also, are we sure that all models that inherits SpamMixin will have a guid?

Yes

File: osf/models/spam.py [hunk 5]
```diff
@@ -266,3 +280,33 @@ def do_check_spam(self, author, author_email, content, request_headers, update=T
...
+            settings.DOMAIN_REGEX,
+            content
+        )
+        if not domains:
+            return False
+
+        from osf.models import NotableDomain
+        if confirm_spam and NotableDomain.has_spam_domain(content):
+            self.confirm_spam(save=True)
+        else:
```
Comment (jwalz) [inline osf/models/spam.py:295]: why is this an `else`? We can't both spam something based on domains *and* add new references?

File: osf/models/spam.py [hunk 6]
```diff
@@ -266,3 +280,33 @@ def do_check_spam(self, author, author_email, content, request_headers, update=T
...
+            self.add_domains_to_moderation_queue(domains)
+
+        return True
+
+    def add_domains_to_moderation_queue(self, domains):
+        from osf.models import NotableDomain, DomainReference
+        for domain in domains:
+            domain = urlparse(domain)
+            notable_domain, created = NotableDomain.objects.get_or_create(
```
Comment (jwalz) [inline osf/models/spam.py:304]: if we pass the `confirm_spam` flag through to this function, we can check the status of the domain here and optionally mark spam at the end of the loop in a single pass instead of calling the status function and having to do the potentially expensive `__in` query above.

File: osf/models/notable_domain.py [hunk 7]
```diff
@@ -40,6 +42,18 @@ def __repr__(self):
     def __str__(self):
         return repr(self)
 
+    @staticmethod
+    def has_spam_domain(content):
```
Comment (jwalz) [inline osf/models/notable_domain.py:46]: As stated elsewhere, I don't think we need this, but if we were to keep it, why pass the content instead of just the discovered domains?

File: osf/models/spam.py [hunk 8]
```diff
@@ -266,3 +280,33 @@ def do_check_spam(self, author, author_email, content, request_headers, update=T
                 self.spam_data['who_flagged'] = 'oopspam'
                 self.spam_data['oopspam_data'] = oopspam_details
         return akismet_is_spam or oopspam_is_spam
+
+    def moderate_domains(self, content, confirm_spam=False):
+        domains = re.findall(
```
Comment (jwalz) [inline osf/models/spam.py:285]: Wrap this in a set, please (ideally using groups and `finditer` as shown in the comment about the regex)

File: osf/models/spam.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jwalz) [reply osf/models/spam.py]: We have no guarantee that classes inheriting from Spam/SpamOverrideMixin also inherit from GuidMixin, but for now, they all do

**Review by jwalz — CHANGES_REQUESTED** [review]
Second pass looks much much better. Sorry for any outdated comments I failed to remove.

File: osf/models/spam.py (hunk 5 above)
Comment (Johnetordoff) [reply osf/models/spam.py:295]: Good point, I will change it so it will capture new all domains and mark them as unknown even if only a single domain is spam.

File: osf/models/notable_domain.py [hunk 9]
```diff
@@ -40,6 +45,56 @@ def __repr__(self):
...
+        return True
+
+    @staticmethod
+    def add_domains_to_moderation_queue(obj):
+        fields_to_query = []
+        spam_fields = list(obj.SPAM_CHECK_FIELDS)
+        for field in spam_fields:
+            fields_to_query += [models.Q(**{f'{field}__regex': settings.DOMAIN_REGEX})]
+        query = reduce(operator.or_, fields_to_query)
+        raw_field_content = str(obj.__class__.objects.filter(query).filter(id=obj.id).values(*spam_fields))
```
[the remark this answers is not in the document]
Comment (Johnetordoff) [reply osf/models/notable_domain.py]: Initially I wanted the code to be isolated from the `SpamMixin`, but obviously it makes more sense seeing as the check only takes place in `do_spam_check`. But since the spam content is always passed to `do_spam_check` it is no longer necessary.

File: osf/external/spam/tasks.py [hunk 10]
```diff
@@ -0,0 +1,25 @@
+import re
+from framework.celery_tasks import app as celery_app
+from website.settings import DOMAIN_REGEX
+from django.contrib.contenttypes.models import ContentType
+
+
+@celery_app.task()
+def check_resource_for_domains(guid, content):
```
Comment (jwalz) [inline osf/external/spam/tasks.py:8]: Ack, this is based on an outdated comment from your first PR pass. If you and @adlius prefer the way you had it before (with more logic in the SpamMixin), I'm find with that. You'd largely reproduced this, just with the spam-check separated out into its own function instead of in the main loop.

File: website/settings/defaults.py [hunk 11]
```diff
@@ -2125,3 +2126,4 @@ def from_node_usage(cls,  usage_bytes, private_limit=None, public_limit=None):
 CAS_LOG_LEVEL = 3  # ERROR
 
 PREPRINT_METRICS_START_DATE = datetime.datetime(2019, 1, 1)
+DOMAIN_REGEX = re.compile('(?P<protocol>\w+://)?(?P<www>www\.)?(?P<domain>[\w-]+\.\w+)(?P<path>/\w*)?')
```
Comment (aaxelb) [inline website/settings/defaults.py:2129]: drive-by: putting a regex in settings smells wrong -- might fit better as a constant defined close to the logic that uses it

File: osf/models/spam.py [hunk 12]
```diff
@@ -221,6 +224,13 @@ def do_check_spam(self, author, author_email, content, request_headers, update=T
         if self.is_spammy:
             return True
 
+        enqueue_task(
```
Comment (jwalz) [inline osf/models/spam.py:227]: What we learned from the `sync_doi_metadata` command is that `enqueue_task` is apparently annoying and doesn't always do what we want. This is _probably_ fine, since this will almost always run in a request context, but we should just default to <task>.apply_async so that we don't have a bunch of different ways that we're running things asynchronously.

For an example, see: https://github.com/CenterForOpenScience/osf.io/blob/develop/osf/management/commands/sync_doi_metadata.py#L36

Comment (jwalz) [inline osf/models/spam.py:227]: Wondering if this should be configurable via settings since so many of our SPAM things are. Thoughts @mfraezz?

File: osf_tests/test_notable_domains.py [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+from osf.models import (
+    NotableDomain,
+    SpamStatus
+)
+from osf.external.spam.tasks import check_resource_for_domains
+from framework.celery_tasks.handlers import enqueue_task
+
+
+@pytest.mark.django_db
+class TestNotableDomain:
```
Comment (jwalz) [inline osf_tests/test_notable_domains.py:19]: Please do a more realistic test of this via the `check_spam` call with domains in SPAM_CHECK_FIELDS (both node and preprint). Also, please confirm that the regex/parsing work as expected by having multiple domains (and multiple entries for the same domain). Also, interactions with `HAM`ed resources/users.

File: osf/models/preprint.py [hunk 14]
```diff
@@ -645,6 +639,14 @@ def save(self, *args, **kwargs):
 
         ret = super(Preprint, self).save(*args, **kwargs)
 
+        if saved_fields and (not settings.SPAM_CHECK_PUBLIC_ONLY or self.verified_publishable):
```
Comment (Johnetordoff) [inline osf/models/preprint.py:642]: Have to put this after the `save()`

File: osf_tests/test_notable_domains.py [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+        NotableDomain.objects.get(
+            domain=spam_domain.netloc,
+            note=NotableDomain.Note.EXCLUDE_FROM_ACCOUNT_CREATION_AND_CONTENT
+        )
+        obj.reload()
+        assert obj.spam_status == SpamStatus.SPAM
+
+    @pytest.mark.parametrize('factory', [NodeFactory, RegistrationFactory, PreprintFactory])
+    def test_spam_check(self, app, factory, spam_domain, marked_as_spam_domain, marked_as_ham_domain):
+        with app.test_request_context(headers={
```
Comment (Johnetordoff) [inline osf_tests/test_notable_domains.py:90]: Spam checks require a request context.

File: osf_tests/test_notable_domains.py [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+            )
+        )
+        obj.reload()
+        NotableDomain.objects.get(
+            domain=spam_domain.netloc,
+            note=NotableDomain.Note.UNKNOWN
+        )
+        obj.reload()
+        assert obj.spam_status == SpamStatus.UNKNOWN
+
```
Comment (jwalz) [inline osf_tests/test_notable_domains.py:65]: Please add a test to ensure that the `DomainReferences` are created

File: osf_tests/test_notable_domains.py [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+from osf.utils.workflows import DefaultStates
+
+from website import settings
+
+
+@pytest.mark.django_db
+class TestNotableDomain:
+
+    @pytest.fixture()
+    def node(self):
```
Comment (jwalz) [inline osf_tests/test_notable_domains.py:27]: This fixture is unused

**Review by jwalz — COMMENTED** [review]
Good catch on the Preprint spam check logic! Just a few test nits

**Comment by adlius** [comment]
One question: how do we deal with the users creating contents containing spammy domains?

File: osf/external/spam/tasks.py [hunk 18]
```diff
@@ -0,0 +1,26 @@
+import re
```
Comment (jwalz) [inline osf/external/spam/tasks.py:1]: nit: not sure how this classifies as "external"

Comment (Johnetordoff) [reply osf/external/spam/tasks.py:1]: The idea is that all the spam services, (oopspam, askismet etc) should be together and they all are external. I didn't expect a spam service to be an internal service.

Comment (jwalz) [reply osf/external/spam/tasks.py:1]: But none of the existing spam things are in this location. You created an `external/spam` directory *just for* an internal task. Can you put this in something like `osf/utils/spam.py` instead?
